tortoise/model/core.go

```diff
@@ -0,0 +1,200 @@
+package model
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+
+	"github.com/spacemeshos/go-spacemesh/activation"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/tortoise"
+)
+
+const (
+	layerSize = 50
+	units     = 10
+)
+
+func newCore(rng *rand.Rand, id string, logger log.Log) *core {
+	db := sql.InMemory()
+	mdb, err := mesh.NewPersistentMeshDB(db, logger)
+	if err != nil {
+		panic(err)
+	}
+	c := &core{
+		id:      id,
+		logger:  logger,
+		rng:     rng,
+		meshdb:  mdb,
+		atxdb:   activation.NewDB(db, nil, nil, types.GetLayersPerEpoch(), types.ATXID{1}, nil, logger),
+		beacons: newBeaconStore(),
+		units:   units,
+		signer:  signing.NewEdSignerFromRand(rng),
+	}
+	cfg := tortoise.DefaultConfig()
+	cfg.LayerSize = layerSize
+	c.tortoise = tortoise.New(c.meshdb, c.atxdb, c.beacons, updater{c.meshdb},
+		tortoise.WithLogger(logger.Named("trtl")),
+		tortoise.WithConfig(cfg),
+	)
+	return c
+}
+
+// core state machine.
+type core struct {
+	id     string
+	logger log.Log
+	rng    *rand.Rand
+
+	meshdb   *mesh.DB
+	atxdb    *activation.DB
+	beacons  *beaconStore
+	tortoise *tortoise.Tortoise
+
+	// generated on setup
+	units  uint32
+	signer signing.Signer
+
+	// set in the first layer of each epoch
+	refBallot     *types.BallotID
+	weight        uint64
+	eligibilities uint32
+
+	// set at the end of the epoch (MessageLayerEnd for the last layer of the epoch)
+	atx types.ATXID
+}
+
+// OnMessage receive blocks, atx, input vector, beacon, coinflip and store them.
+// Generate atx at the end of each epoch.
+// Generate block at the start of every layer.
+func (c *core) OnMessage(m Messenger, event Message) {
+	switch ev := event.(type) {
+	case MessageLayerStart:
+		// TODO(dshulyak) produce ballot according to eligibilities
+		if !ev.LayerID.After(types.GetEffectiveGenesis()) {
+			return
+		}
+		if c.refBallot == nil {
+			total, _, err := c.atxdb.GetEpochWeight(ev.LayerID.GetEpoch())
+			if err != nil {
+				panic(err)
+			}
+			c.eligibilities = max(uint32(c.weight*50/total), 1)
```

Comment:
what's 50?

Reply:
hardcode for the number of ballots, will change it to the consntant
i left a todo on top to enable eligibility computation in the model

Reply:
make it a named constant like `numBallots` bellow the TODO?